src/zenml/orchestrators/step_run_utils.py

```diff
@@ -0,0 +1,466 @@
+    def create_request(self, invocation_id: str) -> StepRunRequest:
+        """Create a step run request.
+
+        This will only create a request with basic information and will not yet
+        compute information like the cache key and inputs. This is separated
+        into a different method `populate_request(...)` that might raise
+        exceptions while trying to compute this information.
+
+        Args:
+            invocation_id: The invocation ID for which to create the request.
+
+        Returns:
+            The step run request.
+        """
+        return StepRunRequest(
+            name=invocation_id,
+            pipeline_run_id=self.pipeline_run.id,
+            deployment=self.deployment.id,
+            status=ExecutionStatus.RUNNING,
+            start_time=datetime.utcnow(),
+            user=Client().active_user.id,
+            workspace=Client().active_workspace.id,
+        )
+
+    def populate_request(self, request: StepRunRequest) -> None:
+        """Populate a step run request with additional information.
+
+        Args:
+            request: The request to populate.
+        """
+        step = self.deployment.step_configurations[request.name]
+
+        input_artifacts, parent_step_ids = input_utils.resolve_step_inputs(
+            step=step,
+            pipeline_run=self.pipeline_run,
+        )
+        input_artifact_ids = {
+            input_name: artifact.id
+            for input_name, artifact in input_artifacts.items()
+        }
+        request.inputs = input_artifact_ids
+        request.parent_step_ids = parent_step_ids
+
+        cache_key = cache_utils.generate_cache_key(
+            step=step,
+            input_artifact_ids=input_artifact_ids,
+            artifact_store=self.stack.artifact_store,
+            workspace_id=Client().active_workspace.id,
+        )
+        request.cache_key = cache_key
+
+        # TODO: Doing this here means this will always fail when running a
+        # template, as the step dependencies are not installed in that case.
+        # In this case, should we just not store the docstring/source code?
+        # Or should we somehow already include that in the deployment? That
+        # would make the deployment object even bigger than it already is
+        # though, as it would contain the code for all steps of the pipeline.
+        (
+            docstring,
+            source_code,
+        ) = self._get_docstring_and_source_code(step=step)
+
+        request.docstring = docstring
+        request.source_code = source_code
+
+        cache_enabled = utils.is_setting_enabled(
+            is_enabled_on_step=step.config.enable_cache,
+            is_enabled_on_pipeline=self.deployment.pipeline_configuration.enable_cache,
+        )
+
+        if cache_enabled:
+            if cached_step_run := cache_utils.get_cached_step_run(
+                cache_key=cache_key
+            ):
+                # TODO: If the step is cached, do we also want to include
```

Comment:
@avishniakov

Reply:
For this I think that using the inputs of the cached step seems better than our current solution.
We already treat the outputs in a way that we reuse the outputs of the cached steps, including dynamic artifacts. If the new step execution would have behaved exactly the same as the cached step (which is sort of the idea of the entire caching mechanism), it would have also dynamically loaded these inputs.

Reply:
Agree, good to have them here